File: wk4/baylor-4238/solution.py
# Baylor 4238 Solution
# see README.md
#
# Jordan Scales <scalesjordan@gmail.com>
# for CS370 (Spring 2014)
# 25 Mar 2014

from math import ceil

def main():
    cases = int(raw_input())

    for i in range(cases):
        # Read in the number of blocks
        n_blocks = int(raw_input())
        block_stream = []

        # Read in the blocks (8 per line)
        for j in range(int(ceil(n_blocks / 8.0))):
            block_stream += raw_input().split(' ')

        blocks = []
        for item in block_stream:
            # Append a tuple
            # Consisting of the item split by commas, converted to ints
            blocks.append(tuple(map(int, item.split(','))))

        # A map of neighbor counts
        neighborMap = {}

        # Add the first block
        neighborMap[blocks[0]] = 0

        # Index to keep track of a block that doesn't share a face with
        # another block
        badBlockIndex = 2
        badBlockFound = False

        for (x, y, z) in blocks[1:]:
            block = (x, y, z)
            neighborMap[block] = 0

            # Add each block to the neighborMap
            for candidate in [(x-1, y, z), (x+1, y, z), \
                              (x, y-1, z), (x, y+1, z), \
                              (x, y, z-1), (x, y, z+1)]:
                if candidate in neighborMap:
                    # A neighbor is in the map!
                    # Increase the count for both our neighbor and ourselves
                    neighborMap[block] += 1
                    neighborMap[candidate] += 1

            if neighborMap[block] == 0:
                # No neighbor :(
                badBlockFound = True
                break

            # Increase the index of the "Bad Block"
            badBlockIndex += 1

        if badBlockFound:
            print '%d NO %d' % (i + 1, badBlockIndex)
        else:
            # Calculate the total surface area
            totalSurfaceArea = 0
            for entry in neighborMap:
                totalSurfaceArea += 6 - neighborMap[entry]

            print '%d %d' % (i + 1, totalSurfaceArea)

if __name__ == '__main__':
    main()

File: wk4/baylor-4238/solution.c
/**
 * Baylor 4238 Solution
 * see README.md
 *
 * Natalie Barillaro, Jordan Scales
 * for CS370 (Spring 2014)
 * 30 Mar 2014
 */

#include <stdio.h>
#include <stdlib.h>

int main(int argc, char **argv) {
    int i, j, k, nPoints, cases, sArea;
    int x, y, z;
    int badBlockIndex, goodBlockFound;
 
    /* Read in the number of cases */
    scanf("%d", &cases);

    for (i = 0; i < cases; i++) {
      /* Read in the number of points */
      scanf("%d", &nPoints);

      /* Allocate space for the array:
         Size of Number of Points  */
      int ***points;
      points = (int***)malloc(nPoints * sizeof(int**));
      for (j = 0; j < nPoints; j++) {
        points[j] = (int**)malloc(nPoints * sizeof(int*));
        for (k = 0; k < nPoints; k++) {
          points[j][k] = (int*)malloc(nPoints * sizeof(int));
        }
      }

      /* The surface area of one block is 6 */
      sArea = 6;

      /* Index to keep track of a block that doesn't
         share a face with another block */
      badBlockIndex = 2;

      /* Store first point */
      scanf("%d,%d,%d", &x, &y, &z);
      points[x][y][z] = 1;

      /* Load the points */
      for (j = 1; j < nPoints; j++) {
        scanf("%d,%d,%d", &x, &y, &z);
        points[x][y][z] = 1;

        sArea += 6;

        goodBlockFound = 0;

        /* Check for neighbors - is there a better way to do this?
           Subtract 2 to surface area for every neighbor */
        if (points[x+1][y][z] == 1) {
          goodBlockFound = 1;
          sArea -= 2;
        }
        if (points[x-1][y][z] == 1) {
          goodBlockFound = 1;
          sArea -= 2;
        }
        if (points[x][y+1][z] == 1) {
          goodBlockFound = 1;
          sArea -= 2;
        }
        if (points[x][y-1][z] == 1) {
          goodBlockFound = 1;
          sArea -= 2;
        }
        if (points[x][y][z+1] == 1) {
          goodBlockFound = 1;
          sArea -= 2;
        }
        if (points[x][y][z-1] == 1) {
          goodBlockFound = 1;
          sArea -= 2;
        }
        if (!goodBlockFound)
          break;

        /* Increase badBlockIndex */
        badBlockIndex++;
      }

      if (goodBlockFound) {
        printf("%d %d", i+1, sArea);
      }

      else {
        printf("%d NO %d", i+1, badBlockIndex);
      }

      printf("\n");

    }

    /* free points (and each point) */
}